packages/nextjs/src/config/wrappers/wrapperUtils.ts

```diff
@@ -1,12 +1,119 @@
+}
+
+/**
+ * Wraps a function that potentially throws. If it does, the error is passed to `captureException` and retrhrown.
+ */
+export function withErrorInstrumentation<F extends (...args: any[]) => any>(
+  origFunction: F,
+): (...params: Parameters<F>) => ReturnType<F> {
+  return function (this: unknown, ...origFunctionArguments: Parameters<F>): ReturnType<F> {
+    const potentialPromiseResult = origFunction.call(this, ...origFunctionArguments);
+    // We do this instead of await so we do not change the method signature of the passed function from `() => unknown` to `() => Promise<unknown>`
+    Promise.resolve(potentialPromiseResult).catch(err => {
+      // TODO: Extract error logic from `withSentry` in here or create a new wrapper with said logic or something like that.
+      captureException(err);
+    });
+    return potentialPromiseResult;
+  };
+}
+
+/**
+ * Calls a server-side data fetching function (that takes a `req` and `res` object in its context) with tracing
+ * instrumentation. A transaction will be created for the incoming request (if it doesn't already exist) in addition to
+ * a span for the wrapped data fetching function.
+ *
+ * All of the above happens in an isolated domain, meaning all thrown errors will be associated with the correct span.
+ *
+ * @param origFunction The data fetching method to call.
+ * @param origFunctionArguments The arguments to call the data fetching method with.
+ * @param req The data fetching function's request object.
+ * @param res The data fetching function's response object.
+ * @param options Options providing details for the created transaction and span.
+ * @returns what the data fetching method call returned.
+ */
+export function callServerSideDataFetcherWithTracingInstrumentation<F extends (...args: any[]) => Promise<any> | any>(
+  origFunction: F,
+  origFunctionArguments: Parameters<F>,
+  req: IncomingMessage,
+  res: ServerResponse,
+  options: {
+    parameterizedRoute: string;
+    functionName: string;
+  },
+): Promise<ReturnType<F>> {
+  return domain.create().bind(async () => {
+    let requestTransaction: Transaction | undefined = getTransactionFromRequest(req);
+
+    if (requestTransaction === undefined) {
+      // TODO: Extract trace data from `req` object (trace and baggage headers) and attach it to transaction
+
+      const newTransaction = startTransaction({
+        op: 'nextjs.data',
+        name: options.parameterizedRoute,
+        metadata: {
+          source: 'route',
+        },
+      });
+
+      requestTransaction = newTransaction;
+      autoEndTransactionOnResponseEnd(newTransaction, res);
+      setTransactionOnRequest(newTransaction, req);
+    }
+
+    const dataFetcherSpan = requestTransaction.startChild({
+      op: 'nextjs.data',
+      description: `${options.functionName} (${options.parameterizedRoute})`,
+    });
+
+    const currentScope = getCurrentHub().getScope();
+    if (currentScope) {
+      currentScope.setSpan(dataFetcherSpan);
```

Comment:
Hmmm... What happens if multiple data-fetching functions for the same request are running at the same time? (Not a current concern, but it [will be one eventually](https://nextjs.org/blog/layouts-rfc#parallel-fetching), and I figure we should future-proof ourselves as much as we can.)

Reply:
I think we're good in this situation. Since we have an individual domain for each of the data fetchers, they also have their own hub/cloned scope, so setting the span should "just work". Let me know if I misunderstand something.

Reply:
> Since we have an individual domain for each of the data fetchers

Yeah, okay, I see what you're saying. I think I'm stuck in the mindset which applies in our other framework integrations, where it's one domain/request. I'm trying to think if there's any reason we'd need to have all of a request's data fetchers live in the same domain. I guess the question is, is there anything which happens to the scope in one which we'd want to be persisted to another? 

Thoughts:

- We don't know the order in which the data fetchers will run, so even if they were all running in the same domain, we can't count on one changing the scope in a way that effects the others, because it might not run first.
- In a way, we're treating the request like a persisted scope, in that we're attaching the transaction to it (the one thing we _do_ need to be the same for all data fetchers).
- I don't think (?) there's anything else we automatically do to the scope which we wouldn't want to lose. 

So I think the answer is no, because we don't know the order in which they'll run, so even if they were in the same domain, we couldn't count on one setting data that the next one could use. The one exception here is the event processor to add request data. That's gotta be attached to whatever scope is active when we call `transaction.finish`. Off-the-cuff idea: In addition to attaching the transaction to the request, we should also store a reference to the request in the transaction metadata. Then we can add an event processor at SDK startup which grabs the request out of `sdkProcessingMetadata` and uses it to populate request data into the event.

Reply:
I spent the night thinking about if there is a **really good reason** for only having one domain across all data fetchers, since having only one domain has its drawbacks with error association. So far I've not come up with anything. I also believe users in any case already have the mental model that data fetchers are more or less isolated and do not share data between them. If bad comes to worst we can explain this behavior in the docs.

As for the request data processor, it turned out to be even simpler. `transaction.finish()` is still called within the individual data fetcher's scope, so adding the event processor inside `callTracedServerSideDataFetcher` just works: 3141759585873690c6a0f8cce8df245d5180788d 👍